core/Test/Tasty/Options.hs

```diff
@@ -143,13 +152,53 @@ mkOptionCLParser mod =
   option parse
     (  long name
     <> help (untag (optionHelp :: Tagged v String))
-    <> mod
+    <> modValidityChecks mod
     )
   where
     name = untag (optionName :: Tagged v String)
     parse = str >>=
       maybe (readerError $ "Could not parse " ++ name) pure <$> parseValue
 
+-- Checks two properties of an option modifier's 'DefaultProp':
+--
+-- 1. The 'value' function has not been used. Doing so interferes with tasty's
+--    ability to read environment variable options, so an error is thrown.
+--
+-- 2. The 'showDefaultWith' function has not been used. Doing so is not
+--    harmful per se, but it will always be ignored in favor of
+--    'showDefaultValue', so a warning is issued.
+--
+-- Ideally, we would be able to inspect the final 'Parser' that was constructed
+-- so that we could issue errors/warnings even if the user did not construct a
+-- Parser by way of 'flagCLParser'/'mkFlagCLParser'/'mkOptionCLParser'.
+-- However, the internal structure of a 'Parser' makes it impossible to
+-- accurately determine whether 'value' or 'showDefaultWith' was used, so
+-- inspecting the 'Mod' is the next best thing. For more information on this
+-- point, see https://github.com/feuerbach/tasty/issues/269.
+modValidityChecks :: forall f v . IsOption v => Mod f v -> Mod f v
+modValidityChecks m@(Mod _ (DefaultProp mbDef mbShowDef) _)
+  | isJust mbDef
+  = error $ unlines
+      [ "ERROR" ++ prov
+      , "Using Options.Applicative.value (or derived functions) in"
+      , "optionCLParser is prohibited, as it interferes with tasty's ability"
+      , "to read environment variable options properly. Moreover, using the"
+      , "value function is unnecessary, as its functionality is subsumed by"
+      , "defaultValue method of IsOption."
+      ]
+  | isJust mbShowDef
+  = trace (unlines
```

Comment:
Using `trace` to display warnings is a bit grimy. But since `modValidityChecks` needs to be run from `mkFlagCLParser`/`mkOptionCLParser` (which are pure functions), I'm not sure how to do better without changing their types to involve `IO`.

Reply:
You can use e.g. a Writer monad to collect the errors, and then print the errors once we reach IO.

Reply:
I'm not sure how that would work in the context of `mkFlagCLParser`/`mkOptionCLParser`, as these exist to implement `optionCLParser`. Unless you are suggesting to change the type of `optionCLParser` to include `Writer`?

Reply:
It doesn't have to be part of `optionCLParser`. It should be a separate post-processing function applied on top of `optionCLParser` inside the `optionParser` function (whose type will change to also return a list of warnings).

Reply:
In other words, we shouldn't rely on the user using mkFlagCLParser or otherwise including modValidityChecks. Instead, we take whatever parser the user provided, and validate and amend it w.r.t. the default values.

Reply:
Unfortunately, it is not possible to perform the validity checks in `modValidityChecks` on a `Parser`. This is because the internal structure of a `Parser a` does not record whether the `a` comes from a command-line argument or a default `value` (see [`mkParser`](https://hackage.haskell.org/package/optparse-applicative-0.15.1.0/docs/src/Options.Applicative.Builder.Internal.html#mkParser)). Moreover, a `Parser` only records the presence of `showDefaultWith` if it is used in combination with `value`. If `showDefaultWith` is used by itself, it will be ignored in the internal structure of the `Parser`'s `OptProperties` (see [`mkProps`](https://hackage.haskell.org/package/optparse-applicative-0.15.1.0/docs/src/Options.Applicative.Builder.Internal.html#mkProps)).

The only way I know how to accurately gauge this information is to look at a `Mod` before it is passed into `mkParser` and loses its internal structure.

Reply:
Ah I see now.

In that case I suggest we test for it "functionally": evaluate the Parser applicative on an empty input and see if it succeeds. In fact, it seems like `evalParser` from `Options.Applicative.Common` does exactly that.

Reply:
Wow, I can't believe I overlooked that! That solves the problem of detecting default values quite nicely. Sadly, I still can't figure out a clean way to detect the presence of `showDefaultWith`, but I think I'll just give up trying to warn about that. (I've added a disclaimer to the Haddocks for `optionCLParser` about this.)